BenchmarkDotNet/BenchmarkRunner.cs

```diff
@@ -295,5 +290,76 @@ private static void AssertBenchmarkMethodIsNotGeneric(MethodInfo methodInfo)
             if (methodInfo.IsGenericMethod)
                 throw new InvalidOperationException($"Benchmark method {methodInfo.Name} is generic.\nGeneric benchmark methods are not supported.");
         }
+
+        /// <summary>
+        /// Given a list of benchmark statistics creates a function to convert 
+        /// raw <see cref="AverageTime"/> measurements to string format so that they
+        /// are shown using uniform time units and align nicely.
+        /// The <see cref="AverageTime"/> measurements are assumed to contain time lengths in nanoseconds.
+        /// </summary>
+        /// <param name="statistics">The list of time-based <see cref="BenchmarkRunReportsStatistic"/>.</param>
+        /// <returns>A function which should be used to convert all <see cref="AverageTime"/> measurements to string.</returns>
+        /// <remarks>
+        /// The measurements are formatted in such a way that they use the same time unit 
+        /// the number of decimals so that they are easily comparable and align nicely.
+        /// 
+        /// Example:
+        /// Consider we have the following raw input where numbers are durations in nanoseconds: 
+        ///     Median=597855, StdErr=485;
+        ///     Median=7643, StdErr=87;
+        /// 
+        /// When using the formatting function, the output will be like this:
+        ///     597.8550 us, 0.0485 us;
+        ///       7.6430 us, 0.0087 us;
+        /// </remarks>
+        public static Func<double, string> GetTimeMeasurementFormattingFunc(IEnumerable<BenchmarkRunReportsStatistic> statistics)
+        {
+            // Find the smallest measurement in the primary statistics, which is the Median.
+            // This will determine the time unit we will use for all measurements.
+            var minRecordedMedian = statistics.Min(stat => stat.AverageTime.Median);
+
+            // Use the largest unit to display the smallest recorded measurement without loss of precision.
+            // TODO: This is duplicated in BenchmarkMeasurementStatistic.ToString(), figure out how to refactor this later?
+            Func<double, string> measurementToString;
+            if (minRecordedMedian < 1000)
```

Comment:
Well surely the original 4 if statements is the simplest possible way and pretty much in line with the existing codebase, isn't it just good enough for now? Aren't we overthinking :)

Besides, you'd still have to implement similar kind of logic in `GetBestTimeUnitFromNanoseconds` right?

Also having this class wouldn't automatically remove the need for `BenchmarkTimeSpan` because it's used in several places. The `BenchmarkTimeSpan.ToString()` probably can be removed, I don't see any direct usages of it.

And to get rid of `BenchmarkTimeSpan` entirely. It would be nice. It's possible to replace it with with `TimeSpan` and a couple of extension methods to get `us` and `ns`. But then everywhere else we use `double` so the use of `TimeSpan` just in one place seems odd. Maybe we could get rid of `BenchmarkTimeSpan` and replace it with raw `double` and just do conversion to seconds (or whatever) directly in one place where it's needed. Or have properties in `BenchmarkRunReportsStatistic` to do `OpsPerSecond` and `AverageNanoseconds` instead of calculating them somewhere else.

You see, there are lots of options and other things we could do, and we also want the formatting done.

How about we do this:

- proceed with the `if` statements as they are for now, if only to make progress with these additions:

    - get rid of `ToString()` in `BenchmarkTimeSpan`
    - move `[BenchmarkTask]` to class level (you seem to care, and I don't have any strong opinion on that).

After this, as a separate PR:

- get rid of `BenchmarkTimeSpan` entirely, or replace it with something else. And we can discuss these changes separately. (And I'm not saying that I'm entirely disagreeing with your idea to have this `MeasurementTimeUnit` class or something similar)

What do you think about this plan.

PS. Do you know what will happen to this discussion if I squash commits and do `push -f`?